src/Proto3/Wire/Decode.hs

```diff
@@ -175,6 +176,8 @@ getKeyVal = do
 getFields :: Get (M.Map FieldNumber (Seq ParsedField))
 getFields = do
     keyvals <- many getKeyVal
+    e <- isEmpty
+    unless e $ fail "Encountered bytes that aren't valid key/value pairs."
```

Comment:
Another option here could be to return the leftovers, and handle the issue with a typed error message elsewhere.